File: ios/TransitKit/Sources/Stores/AlertStore.swift
import Foundation

// MARK: - Alert Store

/// Global GTFS-RT service alerts feed. Polls every 60s and provides indexed lookups
/// by stop_id / route_id. Active-period filtering is applied on the way out so
/// callers see only currently-valid alerts.
@MainActor
@Observable
final class AlertStore {
    // MARK: State
    private(set) var allAlerts: [GtfsRtAlert] = []
    private(set) var lastFetchedAt: Date? = nil

    // MARK: Private
    private var serviceAlertsUrl: String?
    private var pollTask: Task<Void, Never>?

    /// IDs of alerts that were active during the previous fetch. Used by the
    /// consumer to detect *new* alerts (for the in-app toast in Phase E).
    private(set) var previouslyActiveIds: Set<String> = []

    // MARK: - Derived
    /// Currently-active alerts, filtered by activePeriods at "now".
    var activeAlerts: [GtfsRtAlert] {
        let epoch = UInt64(Date().timeIntervalSince1970)
        return allAlerts.filter { $0.isActive(at: epoch) }
    }

    /// Alerts touching a given stop_id.
    func alerts(forStopId stopId: String) -> [GtfsRtAlert] {
        activeAlerts.filter { $0.affectedStopIds.contains(stopId) }
    }

    /// Alerts touching a given route_id.
    func alerts(forRouteId routeId: String) -> [GtfsRtAlert] {
        activeAlerts.filter { $0.affectedRouteIds.contains(routeId) }
    }

    /// True if the given stop has any active alert. O(n) in active alert count.
    func hasAlert(forStopId stopId: String) -> Bool {
        activeAlerts.contains { $0.affectedStopIds.contains(stopId) }
    }

    /// True if the given route has any active alert.
    func hasAlert(forRouteId routeId: String) -> Bool {
        activeAlerts.contains { $0.affectedRouteIds.contains(routeId) }
    }

    // MARK: - Lifecycle

    func configure(serviceAlertsUrl: String?) {
        self.serviceAlertsUrl = serviceAlertsUrl
        startPolling()
    }

    func startPolling() {
        guard serviceAlertsUrl != nil else { return }
        pollTask?.cancel()
        pollTask = Task { [weak self] in
            while !Task.isCancelled {
                await self?.fetch()
                try? await Task.sleep(for: .seconds(60))
            }
        }
    }

    func stopPolling() {
        pollTask?.cancel()
        pollTask = nil
    }

    // MARK: - Fetch

    private func fetch() async {
        guard let urlString = serviceAlertsUrl,
              let url = URL(string: urlString) else { return }
        do {
            let (data, _) = try await URLSession.shared.data(from: url)
            apply(decodeGtfsRtAlerts(from: data))
        } catch {
            // Silent: a transient feed failure shouldn't empty out the UI.
        }
    }

    private func apply(_ alerts: [GtfsRtAlert]) {
        let epoch = UInt64(Date().timeIntervalSince1970)
        previouslyActiveIds = Set(allAlerts
            .filter { $0.isActive(at: epoch) }
            .map(\.id))
        allAlerts = alerts
        lastFetchedAt = Date()
    }
}

File: android/app/src/main/java/com/transitkit/app/data/store/AlertStore.kt
package com.transitkit.app.data.store

import com.transitkit.app.config.OperatorConfig
import com.transitkit.app.data.gtfsrt.GtfsRtFetcher
import com.transitkit.app.data.model.ServiceAlert
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import javax.inject.Inject
import javax.inject.Singleton

private const val POLL_INTERVAL_MS = 60_000L
private const val MAX_POLL_INTERVAL_MS = 300_000L

/**
 * Global GTFS-RT service alerts store. Polls every 60s and exposes O(1) lookups
 * by stop_id / route_id, already filtered by `activePeriods` against "now".
 *
 * Consumers observe `activeAlerts`, `byStopId`, `byRouteId` via StateFlow. The
 * store tracks `previouslyActiveIds` so callers (toast in Phase E) can detect
 * new alerts across poll cycles.
 */
@Singleton
class AlertStore @Inject constructor(
    private val fetcher: GtfsRtFetcher,
    private val config: OperatorConfig,
) {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    private val _allAlerts = MutableStateFlow<List<ServiceAlert>>(emptyList())

    /** All currently-active alerts (activePeriods filter applied). */
    private val _activeAlerts = MutableStateFlow<List<ServiceAlert>>(emptyList())
    val activeAlerts: StateFlow<List<ServiceAlert>> = _activeAlerts.asStateFlow()

    /** active alerts indexed by stop_id. */
    private val _byStopId = MutableStateFlow<Map<String, List<ServiceAlert>>>(emptyMap())
    val byStopId: StateFlow<Map<String, List<ServiceAlert>>> = _byStopId.asStateFlow()

    /** active alerts indexed by route_id. */
    private val _byRouteId = MutableStateFlow<Map<String, List<ServiceAlert>>>(emptyMap())
    val byRouteId: StateFlow<Map<String, List<ServiceAlert>>> = _byRouteId.asStateFlow()

    /** IDs that were active in the previous poll cycle. Used for new-alert detection. */
    private val _previouslyActiveIds = MutableStateFlow<Set<String>>(emptySet())
    val previouslyActiveIds: StateFlow<Set<String>> = _previouslyActiveIds.asStateFlow()

    private var pollingJob: Job? = null
    private var consecutiveErrors = 0

    fun startPolling() {
        val url = config.gtfsRt?.serviceAlertsUrl ?: return
        if (pollingJob?.isActive == true) return

        pollingJob = scope.launch {
            while (isActive) {
                val result = runCatching { fetcher.fetchAlerts(url) }
                if (result.isSuccess) {
                    apply(result.getOrThrow())
                    consecutiveErrors = 0
                } else {
                    consecutiveErrors++
                }
                val backoffMs = minOf(
                    POLL_INTERVAL_MS shl consecutiveErrors.coerceAtMost(3),
                    MAX_POLL_INTERVAL_MS,
                )
                delay(backoffMs)
            }
        }
    }

    fun stopPolling() {
        pollingJob?.cancel()
        pollingJob = null
        consecutiveErrors = 0
    }

    // -------------------------------------------------------------------------
    // Apply

    private fun apply(alerts: List<ServiceAlert>) {
        val epoch = System.currentTimeMillis() / 1000
        val previouslyActive = _activeAlerts.value.map { it.id }.toSet()
        val active = alerts.filter { it.isActive(epoch) }

        _previouslyActiveIds.value = previouslyActive
        _allAlerts.value = alerts
        _activeAlerts.value = active

        val stopIndex = mutableMapOf<String, MutableList<ServiceAlert>>()
        val routeIndex = mutableMapOf<String, MutableList<ServiceAlert>>()
        for (a in active) {
            for (s in a.affectedStopIds) stopIndex.getOrPut(s) { mutableListOf() }.add(a)
            for (r in a.affectedRouteIds) routeIndex.getOrPut(r) { mutableListOf() }.add(a)
        }
        _byStopId.value = stopIndex
        _byRouteId.value = routeIndex
    }
}
